packages/js/product-editor/src/data-form/components/fields/schedule-sale/field.tsx

```diff
@@ -0,0 +1,231 @@
+/**
+ * External dependencies
+ */
+import { DateTimePickerControl } from '@woocommerce/components';
+import { Product } from '@woocommerce/data';
+import { recordEvent } from '@woocommerce/tracks';
+import { ToggleControl } from '@wordpress/components';
+import {
+	createElement,
+	useCallback,
+	useEffect,
+	useState,
+} from '@wordpress/element';
+import { DataFormControlProps } from '@wordpress/dataviews';
+import { __ } from '@wordpress/i18n';
+import moment from 'moment';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore We need this to get the datetime format for the DateTimePickerControl.
+// eslint-disable-next-line @woocommerce/dependency-group
+import { getSettings } from '@wordpress/date';
+
+/**
+ * Internal dependencies
+ */
+import { useProductEdits } from '../../../../hooks/use-product-edits';
+import { useValidation } from '../../../../contexts/validation-context';
+
+export function ScheduleSaleFieldEdit( {
+	data,
+	onChange,
+}: DataFormControlProps< Product > ) {
+	const salePrice = data.sale_price;
+	const dateOnSaleFromGmt = data.date_on_sale_from_gmt;
+	const dateOnSaleToGmt = data.date_on_sale_to_gmt;
+
+	const { hasEdit } = useProductEdits();
+
+	const dateTimeFormat = getSettings().formats.datetime;
+
+	const [ showScheduleSale, setShowScheduleSale ] = useState( false );
+
+	const isSalePriceGreaterThanZero =
+		Number.parseFloat( salePrice || '0' ) > 0;
+
+	const today = moment().startOf( 'minute' ).toISOString();
+
+	const setDateOnSaleFromGmt = useCallback(
+		( value: string ) => {
+			onChange( {
+				date_on_sale_from_gmt: value,
+			} );
+		},
+		[ onChange ]
+	);
+
+	const setDateOnSaleToGmt = useCallback(
+		( value: string ) => {
+			onChange( {
+				date_on_sale_to_gmt: value,
+			} );
+		},
+		[ onChange ]
+	);
+
+	function handleToggleChange( value: boolean ) {
+		recordEvent( 'product_pricing_schedule_sale_toggle_click', {
+			enabled: value,
+		} );
+
+		setShowScheduleSale( value );
+
+		if ( value ) {
+			setDateOnSaleFromGmt( today );
+			setDateOnSaleToGmt( '' );
+		} else {
+			setDateOnSaleFromGmt( '' );
+			setDateOnSaleToGmt( '' );
+		}
+	}
+
+	// Hide and clean date fields if the user manually changes
+	// the sale price to zero or less.
+	useEffect( () => {
+		if ( hasEdit( 'sale_price' ) && ! isSalePriceGreaterThanZero ) {
+			setShowScheduleSale( false );
+			setDateOnSaleFromGmt( '' );
+			setDateOnSaleToGmt( '' );
+		}
+	}, [
+		hasEdit,
```

Comment:
The addition of `hasEdit` in this dependency array causes an infinite loop when updating an existing product. Because `hasEdit` gets regenerated on every render ( its not using a `useCallback` ).